Added Gradle Wrapper support

gmazzo opened the pull request:
### What are you trying to accomplish?
Introduces `Gradle Wrapper` bump support for `gradle` manager

Closes #2223 (issue open since 2018)

Smoke Tests PR: https://github.com/dependabot/smoke-tests/pull/325
GitHub Docs PR: https://github.com/github/docs/pull/39954

### Anything you want to highlight for special attention from reviewers?
Following the same pattern used by `gradle/libs.version.toml` file, we'll only look (and support) for a `gradle/wrapper/gradle-wrapper.properties` in the root of the project.

`distributionSha256Sum` is also supported, and it will only be updated if the original properties file has the property.

> [!NOTE]
> To minimize the risk of the feature and the impact on other smoke tests, I've introduced a new `gradle_wrapper_updater` feature flag for this change.

#### JAR binary and script files support
Updating other the `gradle-wrapper.jar` or the companion shell scripts is done through the new `WrapperUpdater` helper by leveraging the existing `LockfileUpdater`. 
Most of its code was extracted into a new `GradleUpdaterBase` class, which has two abstract methods:
- `target_file?` to tell the given file is meant to be updated by it
- `command_args` to collaborate with the final `gradle` command run

Now both `LockfileUpdater` and the new `WrapperUpdater` inherit from `GradleUpdaterBase`.

> [!NOTE]
> The update will not attempt to use any configuration for the `wrapper` task in the main build script; many of the repos will be broken because they rely on other files or included builds that will take a long time to complete. We limit to copying the wrapper-related files and running a clean, empty build.

### How will you know you've accomplished your goal?
Running against a sample repo:
```sh
bin/docker-dev-shell gradle --rebuild
$ bin/dry-run.rb gradle gmazzo/gradle-codeowners-plugin
```
produces the desired diff:
```
 => bump gradle-wrapper from 8.14.3 to 9.0.0

    ± gradle/wrapper/gradle-wrapper.properties
    ~~~
    --- /tmp/original20250820-3941-jbt37h       2025-08-20 12:30:57.699516752 +0000
    +++ /tmp/updated20250820-3941-peq4i4        2025-08-20 12:30:57.700516752 +0000
    @@ -1,7 +1,7 @@
     distributionBase=GRADLE_USER_HOME
     distributionPath=wrapper/dists
    -distributionSha256Sum=bd71102213493060956ec229d946beee57158dbd89d0e62b91bca0fa2c5f3531
    -distributionUrl=https\://services.gradle.org/distributions/gradle-8.14.3-bin.zip
    +distributionSha256Sum=8fad3d78296ca518113f3d29016617c7f9367dc005f932bd9d93bf45ba46072b
    +distributionUrl=https\://services.gradle.org/distributions/gradle-9.0.0-bin.zip
     networkTimeout=10000
     validateDistributionUrl=true
     zipStoreBase=GRADLE_USER_HOME
    ~~~
    3 insertions (+), 3 deletions (-)
```


### Checklist

<!-- Before requesting review, please ensure that your pull request fulfills the following requirements: -->

- [x] I have run the complete test suite to ensure all tests and linters pass.
- [x] I have thoroughly tested my code changes to ensure they work as expected, including adding additional tests for new functionality.
- [x] I have written clear and descriptive commit messages.
- [x] I have provided a detailed description of the changes in the pull request, including the problem it addresses, how it fixes the problem, and any relevant details about the implementation.
- [x] I have ensured that the code is well-documented and easy to understand.

> [!IMPORTANT]
> Do not squash this PR, so we won't the history of `lockfile_updater.rb` being renamed to `gradle_updater_base.rb` and to attribute the original code to its autor.

yeikel commented on gradle/lib/dependabot/gradle/distributions.rb:
```diff
@@ -0,0 +1,22 @@
+    module Distributions
+      extend T::Sig
+
+      DISTRIBUTIONS_URL = "https://services.gradle.org"
```
Can we detect this from the `gradle-wrapper.properties`? 

In my current configuration, for example, this is pointing to a custom proxy. Example

```properties
distributionBase=GRADLE_USER_HOME
distributionPath=wrapper/dists
distributionUrl=https://example.com/gradle-distributions/gradle-9.0.0-bin.zip
validateDistributionUrl=true
zipStoreBase=GRADLE_USER_HOME
zipStorePath=wrapper/dists
distributionBuildTime=20250731163512+0000
```

gmazzo replied:
This is pretty much intentional, because: 
- The PR is aiming to support the 95% of the cases (if not more) where people use a standard Gradle distribution
- Even if I manage to infer the base URL for a non standard distribution, there won't be any guarantee that in that same server will exist an equivalent listing endpoint https://services.gradle.org/versions/all, or it will honor it's contact 

Therefore the extra complexity to support this edge case won't pay off, at least in the scope of this PR

yeikel reviewed (commented):
Thank you for making this change. It would be a great addition! 

A few additional points regarding the review:

A key aspect of full Gradle wrapper support is regenerating the wrapper scripts (like `gradlew` and related files). If only the property file is updated and the scripts aren’t regenerated, it could cause issues.

In my view, a proper Gradle Wrapper upgrade should:

- Update the distribution property file
- Use the native Gradle binary to regenerate the wrapper scripts
- Commit both changes

For reference, see how the [Update Gradle Wrapper Action](https://github.com/marketplace/actions/update-gradle-wrapper-action) handles this process.

As a side note, considering recent moves to separate ecosystems (such as `npm` and `bum`), it might be worth discussing if this should become its own ecosystem. I understand that gradle and the wrapper are related, but the structure and goal are different. It could be structured either way. 

I’ll leave that decision to the Dependabot team, as there are both advantages and disadvantages.

gmazzo commented:
> Thank you for making this change. It would be a great addition! 
> 
> A few additional points regarding the review:
> 
> A key aspect of full Gradle wrapper support is regenerating the wrapper scripts (like `gradlew` and related files). If only the property file is updated and the scripts aren’t regenerated, it could cause issues.
> 
> In my view, a proper Gradle Wrapper upgrade should:
> 
> - Update the distribution property file
> - Use the native Gradle binary to regenerate the wrapper scripts
> - Commit both changes
> 
> For reference, see how the [Update Gradle Wrapper Action](https://github.com/marketplace/actions/update-gradle-wrapper-action) handles this process.
> 
> As a side note, considering recent moves to separate ecosystems (such as `npm` and `bum`), it might be worth discussing if this should become its own ecosystem. I understand that gradle and the wrapper are related, but the structure and goal are different. Ultimately, however, it will be matter of opinion
> 
> I’ll leave that decision to the Dependabot team, as there are both advantages and disadvantages.

Thanks for the review @yeikel !

Let's wait for the maintainers feedback, but IMHO updating the whole set of scripts as that action does is just against the spirit of the tool itself. Not to mention it will be a technical challenge (or a blocker) to implement properly. 

I dig into this tool a few days ago, so I'm not an expert at all. But based on what I get from experience, it works with a "token replacement" approach.

Implementing a proper script updating will require either to run Gradle or to do reverse engineering to the `wrapper` to mimic its behavior. That's just not the way this tool seems to work. 

If a full update is required/desired, there are others tools for that or you can just do it manually yourself. 

There better "Gradle-native" approaches to update dependencies that does not has the flaws the Dependabot has. This tool just provides a static fast approach that fits in the 90% of the cases (again or even more). The wrapper updater just falls under the same pattern IMHO

> If only the property file is updated and the scripts aren’t regenerated, it could cause issues.

In my experience, this has never happened. The wrapper API has been pretty stable from Gradle earlier versions 

> it might be worth discussing if this should become its own ecosystem.

Well, IMHO I disagree. That Gradle has a different set of things to update, does not implies they are different ecosystems. If your thoughts are based on the Renovate approach, I'm not sure why it was implemented like that, but high-level, Gradle is just one. Its wrapper is just another kind of updatable component.

yeikel commented on gradle/lib/dependabot/gradle/file_parser/distributions_finder.rb:
```diff
@@ -0,0 +1,86 @@
+        extend T::Sig
+        include Dependabot::Gradle::Distributions
+
+        DISTRIBUTION_URL_REGEX = %r{^#{Regexp.escape(DISTRIBUTIONS_URL)}/distributions/gradle-(?<version>[\d.]+)-(?<distribution_type>bin|all)\.zip$} # rubocop:disable Layout/LineLength
```
This may work 

```ruby
DISTRIBUTION_URL_REGEX = %r{
  ^#{Regexp.escape(DISTRIBUTIONS_URL)}/distributions/
  gradle-(?<version>[\d.]+)-
  (?<distribution_type>bin|all)\.zip$
}x
```

gmazzo commented:
> regenerating lock files or running mvn commands to get additional insights from the Maven directly.

Thanks for pointing that out, @yeikel. I just noted there are effective piece of code that run native tools, [like `mvn`](https://github.com/dependabot/dependabot-core/blob/main/maven/lib/dependabot/maven/native_helpers.rb#L17)

I was originally reluctant to propose a solution like that, because running `gradle` requires having `java` installed as well. Which, at this point I'd assume it does if it's running `maven`.

Now I think a complementary step may be added to use `./gradlew wrapper` task instead, falling back to this original approach if it fails. 

Let me see what I can do, or it can be done in a follow-up PR too. I'd like to get feedback from the maintainers about this as well.

gmazzo commented:
> you should make sure that all the tests are passing as that'll be the first feedback you'll get 🙏

I coudn't find documentation on how to work with the smoke tests. They are in a different repo. I have a PR for them as well, but changing `SMOKE_TEST_BRANCH` does not feel correct either 🤔

gmazzo commented:
> * Smoke / e2e (docker, docker, smoke-docker-version-multidir.yaml, 0dffddee7cfa3e9521be1813cd531d4e7047c098) (pull_request)

I rebased the PRs and fixed the lint issues. Everything seems to be green now ;)

gmazzo commented:
Hi @kbukum1, any feedback on this?

gmazzo commented:
Conflicts 🤦

kbukum1 commented on .github/workflows/smoke.yml:
```diff
@@ -14,7 +14,8 @@ concurrency:
   GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
-  SMOKE_TEST_BRANCH: main
+  SMOKE_TEST_REPO: gmazzo/dependabot-smoke-tests
+  SMOKE_TEST_BRANCH: gradle-wrapper-support
```
Just reminder: Needs to be reverted after approval.

gmazzo replied:
> Can we retain the new environment variable SMOKE_TEST_REPO and just revert it to dependabot/smoke-tests? This is useful for other maintainers

That was my original thought on this. Smoke tests won't pass without this, or I'm not sure exactly how to do it. I tried to follow the docs, but it wasn't successful

What about using repo variables instead?
- `${{ vars.SMOKE_TEST_REPO || 'dependabot/smoke-tests' }}`
- `${{ vars.SMOKE_TEST_BRANCH || 'main' }}`

kbukum1 replied:
It is possible to have that but we shouldn't have this in this PR. It can be created in another PR and that PR can be shown as reference?

kbukum1 commented on .github/workflows/smoke.yml:
```diff
@@ -44,7 +45,7 @@ jobs:
 
           # Curl the smoke-test tests directory to get a list of tests to run
-          URL=https://api.github.com/repos/dependabot/smoke-tests/contents/tests?ref=${{ env.SMOKE_TEST_BRANCH }}
+          URL=https://api.github.com/repos/${{ env.SMOKE_TEST_REPO }}/contents/tests?ref=${{ env.SMOKE_TEST_BRANCH }}
```
Just reminder: Needs to be reverted after approval.

gmazzo replied:
@kbukum1 let me know what you think about https://github.com/dependabot/dependabot-core/pull/12891#discussion_r2470542957

kbukum1 replied:
https://github.com/dependabot/dependabot-core/pull/12891#discussion_r2470824385

kbukum1 commented on gradle/lib/dependabot/gradle/update_checker/version_finder.rb:
```diff
@@ -118,6 +120,10 @@ def lowest_security_fix_version_details
         def versions
+          if Distributions.distribution_requirements?(dependency.requirements)
+            return DistributionsFinder.available_versions
+          end
```
There are a few issues here that need to be addressed:

1. **Consolidate version fetching logic**: `Package::PackageDetailsFetcher` is designed to fetch package versions, so the distribution fetching should be moved into `PackageDetailsFetcher` as well. This way we have a single method to fetch all versions.

2. **Add feature flag check**: We need to check `Experiments.enabled?(:gradle_wrapper_updater)` before fetching distributions, just like we did in the file parsing logic.

3. **Include `released_at` field**: The current format is missing the `released_at` field, which means cooldown functionality won't work for Gradle wrapper updates.

Current format:
```ruby
{
  version: version,
  source_url: source_url
}
```

Should match the format used by `PackageDetailsFetcher`:
```ruby
{
  version: version,
  released_at: released_at,
  source_url: source_url
}
```

gmazzo replied:
Point 1 seems to be the most challenging, let me see what I do

gmazzo replied:
Ok, @kbukum1, I think I covered all the feedback. I still kept some of the new logic isolated out from `PackageDetailsFetcher` (in a `DistributionsFetcher` sibling) but it's definitely integrated into it now.

I think this is better because it won't unnecessary pollute the `PackageDetailsFetcher` (which is more Maven related) with this, but let me know what you think ;)

kbukum1 commented on gradle/lib/dependabot/gradle/update_checker/requirements_updater.rb:
```diff
@@ -114,6 +117,36 @@ def update_dynamic_requirement(req_string)
         end
 
+        sig { returns(T::Array[T::Hash[Symbol, T.untyped]]) }
+        def updated_distribution_requirements
```
We need to return false if our feature flag disabled?

gmazzo commented on gradle/lib/dependabot/gradle/file_updater/lockfile_updater.rb:
I did the refactor to avoid duplicating code. But considering (based on your review) that isolating changes has more weight than duplicating a small piece of code, I think it's just better (at least for the scope of this PR) to just create two independent clases, without a common base for now.

I can undertake this once this becomes stable.

gmazzo replied:
Done

gmazzo commented on gradle/lib/dependabot/gradle/file_updater.rb:
```diff
@@ -93,17 +99,28 @@ def update_buildfiles_for_dependency(buildfiles:, dependency:)
+          if existing_file.nil?
+            files << file
+          else
+            files[T.must(files.index(existing_file))] = file
```
This could be now the most "controversial" refactor, so let me explain a bit why.

Before you were just calling `LockfileUpdater.update_lockfiles` on the `buildfile` and collecting its affecting files. That's OK.

Now (and to avoid duplicating the logic, but I'd be fine if that's what you want/need) I did:
- Created a `updated_files` array to collect files from both `LockfileUpdater` and the new `WrapperUpdater` (they will only run if their experiments are enabled)
- Then run the last piece of logic of "affected files" for the resulting collection. `lockfiles.each do |lockfile|` was renamed to `updated_files.each do |file|` just for semantic coherence.
- You may also noted the `buildfiles_processed` map:
  - This was needed because the actual logic will be run for every `requirement` in the `dependency`.
  - For Gradle, we never required more than just, just the `version` bump
  - But in order to fully support Gradle Distribution, I modeled the `dependency` with two requirements:
    - The `version`
    - And the `distributionSHA`
  - The thing is, without change, the first run (for the `version` requirement) of the native tool (`gradle wrapper`) will update both of time, in the `gradle.properties` file.
  - Because of this, a further second run will fail, because no files will be updated. Also a second run will be unnecessary and just a wast of time and resources for the tool.
  - So to overcome this, is what I "keep track" of build files were I run the native tools already, so I won't run them again.

kbukum1 replied:
Thanks for the explanation. I think it is clear now how it is works.

gmazzo commented on gradle/lib/dependabot/gradle/package/package_details_fetcher.rb:
```diff
@@ -83,7 +85,7 @@ def fetch_available_versions
             package_releases << {
               version: Gradle::Version.new(version),
-              released_at: release_date_info.none? ? nil : (release_date_info[version]&.fetch(:release_date) || nil),
+              released_at: info[:released_at] || release_date_info[version]&.fetch(:release_date),
```
This change should be equivalent to the code before, except that now if `released_at` is set directly in `info` (only applies for distributions) it will take priority. This was to avoid duplicating the `release_date_info` structure from Maven packages

gmazzo commented on gradle/spec/dependabot/gradle/package/distributions_fetcher_spec.rb:
```diff
@@ -0,0 +1,648 @@
+            version: "0.7",
+            build_time: "20090720085013+0200"
+          }
+        ].sort_by { |v| v[:version] }
```
Just pushed, option 2

yeikel commented on gradle/lib/dependabot/gradle/file_updater/wrapper_updater.rb:
```diff
@@ -0,0 +1,161 @@
+            cwd = File.join(temp_dir, base_path(build_file))
+
+            # Create gradle.properties file with proxy settings
+            # Would prefer to use command line arguments, but they don't work.
```
What command line arguments don't work?

gmazzo replied:
This is actually coming from the original `lockfile_updater`:
https://github.com/dependabot/dependabot-core/blob/main/gradle/lib/dependabot/gradle/file_updater/lockfile_updater.rb#L37

Since we agreed duplicating the code, I also copied this.

I did researched a bit on this, and it seems that if you pass the `systemProp.http.proxyHost=` related params as environment variables, it's not working for the `gradlew` download logic (the logic that downloads the gradle distribution binary). 
So the workaround was to dump it in a `gradle.properties` file, that seems to be supported

yeikel commented on gradle/spec/dependabot/gradle/file_parser/distributions_finder_spec.rb:
```diff
@@ -0,0 +1,88 @@
+      %w(https://services.gradle.org/distributions/gradle-9.1.0-rc-1-bin.zip 9.1.0-rc-1),
+      %w(https://my.company.org/dists/my-gradle-9.1.0-rc-1-bin.zip 9.1.0-rc-1),
+      %w(https://my.company.org/dists/my-gradle-9.1.0-rc-2-bin-customized.zip 9.1.0-rc-2),
+      %w(https://my.company.org/dists/my-gradle-8.1.tar 8.1),
```
Thank you for this!

yeikel commented on gradle/lib/dependabot/gradle/file_parser.rb:
```diff
@@ -59,6 +60,11 @@ def parse
         script_plugin_files.each do |plugin_file|
           dependency_set += buildfile_dependencies(plugin_file)
         end
+        if Experiments.enabled?(:gradle_wrapper_updater)
```
@kbukum1 For my own understanding, what's the plan for this? How long do we normally run these as experiments and what % of the base users will see this running?

kbukum1 commented:
@gmazzo , @yeikel 

I did deploy the change. If you have a repository you want to test current maven update you can check. In mean time I am also doing some checks, if everything goes well, we will merge the change. Currently I am testing if the change causes any problem to existing workflow behind feature flag.

kbukum1 commented:
I did currently tested for current update process and it worked properly. I am going to merge the changes. 

@gmazzo , if you have repositories you can test, I can active feature flag for this repositories and you will be able to test your changes.  I can active it for the one you shared, `gmazzo/gradle-codeowners-plugin`

As for how to test on your repository, you just need currently either make changes in `dependabot.yml` or rerun the update process through `Insights -> Depencency Graph -> Dependabot -> Rerun Update . However if you run current update without enabling feature flag, it should use the existing update process instead without wrapper update support. If this repository only created for wrapper update, after I enable the feature flag, you can run update for the repository.

kbukum1 commented:
@gmazzo 
I think it is ok. You can update your `dependaboy.yml` to retrigger. There is a certain bug happening when triggering manually. I will check it later but not related to your changes. You can add space, comment or change anything in dependabot.yml which will retrigger full process.

yeikel commented:
Thank you for clarifying @kbukum1 

I think that the best benchmarks are real and popular gradle projects that use dependabot

Here are a couple of forks we can try: 

https://github.com/yeikel/kafka-ui
https://github.com/yeikel/testcontainers-java
https://github.com/yeikel/git-machete-intellij-plugin/tree/develop

gmazzo commented:
@kbukum1, the GH UI reports this error:
<img width="596" height="143" alt="image" src="https://github.com/user-attachments/assets/b754ac61-6202-4f51-9982-7375f1d8eb18" />

But I couldn't find anything on from where it comes from on `dependabot`'s public repos.

The run fails when run inside GH with
```
The request contains invalid or unauthorized changes
```
https://github.com/gmazzo/gradle-codeowners-plugin/actions/runs/19097588727/job/54561307423#step:3:708

That message, I manage to map it to `dependabot-core`'s logic, but it's just "error handling" when calling the client api `"#{base_url}/update_jobs/#{job_id}/create_pull_request"`:
https://github.com/dependabot/dependabot-core/blob/main/updater/lib/dependabot/api_client.rb#L511

But that's where I lose track of the logic, is possible this an internal GH security validation that is not public code?

kbukum1 commented:
@yeikel , 

```
Successfully enabled gradle_wrapper_updater on yeikel/kafka-ui
Repository yeikel/testcontainers-java not found.
Successfully enabled gradle_wrapper_updater on yeikel/git-machete-intellij-plugin/tree/develop
```

yeikel commented:
> https://github.com/yeikel/testcontainers-java

It should be enabled now. Sorry for the confusion!

yeikel commented:
In one of the repositories, I am seeing the following failure 

> 2025/11/07 21:26:25 INFO <job_1146444339> Total execution time: 4.81 seconds
> updater | Failed to update files: To honour the JVM settings for this build a single-use Daemon process will be forked. For more on this, please refer to https://docs.gradle.org/8.14.2/userguide/gradle_daemon.html#sec:disabling_the_daemon in the Gradle documentation.
> Daemon will be stopped at the end of the build 
> FAILURE: Build failed with an exception.
> * What went wrong:
> Project directory 'dependabot_tmp_dir/gradle' is not part of the build defined by settings file 'dependabot_tmp_dir/settings.gradle'. If this is an unrelated build, it must have its own settings file.
> * Try:
> > Run with --info or --debug option to get more log output.
> > Run with --scan to get full insights.

This seems to be a consequence of moving the build files around instead of running it from the root directly of the project. We'll have to find a workaround for that 

Full logs: https://github.com/yeikel/kafka-ui/actions/runs/19181649304/job/54839606560

gmazzo commented:
FYI @kbukum1 @yeikel, I opened a new PR fixing a couple of issues (one critical) I noted from the tests runs:
https://github.com/dependabot/dependabot-core/pull/13501
